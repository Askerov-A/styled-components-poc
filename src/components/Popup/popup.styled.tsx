import styled, { css, keyframes } from "styled-components";

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

export const SPopupWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgba(0, 0, 0, 0.4);
  transition: all 0.3s ease;
  animation: ${fadeIn} 500ms;
`;

interface PopupContentWrapperProps {
  readonly width?: number;
}

export const SPopupContentWrapper = styled.div<PopupContentWrapperProps>`
  position: relative;
  width: 80%;
  margin: 15% auto;
  padding: 20px;
  background-color: #fefefe;
  border-radius: 4px;
  width: 100%;
  ${({ width }) =>
    width &&
    css`
      width: ${width}px;
    `};
`;
export const SPopupContent = styled.div``;
export const SPopupHeader = styled.div``;
export const SPopupTitle = styled.div``;
export const SPopupButton = styled.button`
  position: absolute;
  top: 16px;
  right: 16px;
  color: #000;
  background: none;
  border: none;
  cursor: pointer;
`;
