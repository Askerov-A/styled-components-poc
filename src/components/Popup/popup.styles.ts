import styled, { css, keyframes } from "styled-components";

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

export const PopupWrapper = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgba(0, 0, 0, 0.4);
  transition: all 0.3s ease;
  animation: ${fadeIn} 500ms;
`;

interface PopupContentWrapperProps {
  readonly width?: number;
}

export const PopupContentWrapper = styled.div<PopupContentWrapperProps>`
  position: relative;
  width: 80%;
  margin: 15% auto;
  padding: 20px;
  background-color: #fefefe;
  border-radius: 4px;
  ${({ width }) => {
    if (width) {
      return css`
        width: ${width}px;
      `;
    }

    return css`
      width: 100%;
    `;
  }}
`;
export const PopupContent = styled.div``;

export const PopupHeader = styled.div``;
export const PopupTitle = styled.div``;
export const PopupButton = styled.button`
  position: absolute;
  top: 16px;
  right: 16px;
  color: #000;
  background: none;
  border: none;
  cursor: pointer;
`;
