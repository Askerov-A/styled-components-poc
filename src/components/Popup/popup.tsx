import React, { useRef, useEffect } from "react";
import ReactDOM from "react-dom";

import {
  SPopupWrapper,
  SPopupContentWrapper,
  SPopupHeader,
  SPopupTitle,
  SPopupButton,
  SPopupContent,
} from "./popup.styled";

interface IPopupProps {
  isShowed: boolean;
  onClose: () => void;
  width?: number;
  headerText?: string;
}

const useOutsideClickHandlerHook = (
  wrapperRef: React.RefObject<HTMLDivElement>,
  onClose: () => void
) => {
  useEffect(() => {
    function handleClickOutside(event: React.MouseEvent) {
      if (
        wrapperRef &&
        wrapperRef.current &&
        !wrapperRef.current.contains(event.target as Node)
      ) {
        onClose();
      }
    }
    document.addEventListener("mousedown", (((e: React.MouseEvent) => {
      handleClickOutside(e);
    }) as unknown) as EventListener);
    return () => {
      document.removeEventListener("mousedown", (((e: React.MouseEvent) => {
        handleClickOutside(e);
      }) as unknown) as EventListener);
    };
  }, [onClose, wrapperRef]);
};

const Popup = ({
  isShowed,
  onClose,
  headerText,
  children,
  width,
}: React.PropsWithChildren<IPopupProps>): JSX.Element | null => {
  const wrapperRef = useRef<HTMLDivElement>(null);
  const domNode: HTMLElement = document.createElement("div");
  domNode.setAttribute("id", "Popup-root");

  const onCloseHandler = () => {
    domNode && document.body.removeChild(domNode);
    onClose();
  };
  useOutsideClickHandlerHook(wrapperRef, onCloseHandler);
  useEffect(() => {
    domNode && isShowed && document.body.appendChild(domNode);
  }, [domNode, isShowed]);
  return (
    <>
      {ReactDOM.createPortal(
        <SPopupWrapper role="presentation">
          <SPopupContentWrapper ref={wrapperRef} width={width}>
            {headerText && (
              <SPopupHeader>
                <SPopupTitle>{headerText}</SPopupTitle>
                <SPopupButton
                  aria-label="Close popup"
                  type="button"
                  tabIndex={0}
                  onClick={onCloseHandler}
                  onKeyDown={onCloseHandler}
                  onKeyPress={onCloseHandler}
                >
                  X
                </SPopupButton>
              </SPopupHeader>
            )}
            <SPopupContent>{children}</SPopupContent>
          </SPopupContentWrapper>
        </SPopupWrapper>,
        domNode
      )}
    </>
  );
};

export default Popup;
