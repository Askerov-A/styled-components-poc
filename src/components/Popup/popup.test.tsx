import { render, screen, fireEvent } from "@testing-library/react";

import Popup from "./popup";

const PROPS = {
  isShowed: true,
  onClose: jest.fn(),
  headerText: "Test header text",
};

describe("Popup Component", () => {
  it("should render correctly", () => {
    const wrapper = render(<Popup {...PROPS}>Test</Popup>);

    expect(wrapper.getByRole("presentation")).toMatchSnapshot();
  });

  it("should find header text", () => {
    const { getByText } = render(<Popup {...PROPS} />);
    expect(getByText(PROPS.headerText)).toBeTruthy();
  });

  it("should call event on click", () => {
    render(<Popup {...PROPS}>Test</Popup>);
    fireEvent.click(screen.getByRole("button"));
    expect(PROPS.onClose).toHaveBeenCalledTimes(1);
  });

  it("should call event on key Press", () => {
    render(<Popup {...PROPS} />);
    fireEvent.keyDown(screen.getByRole("button"), { key: "Enter", code: 13 });
    expect(PROPS.onClose).toHaveBeenCalledTimes(1);
  });
});
