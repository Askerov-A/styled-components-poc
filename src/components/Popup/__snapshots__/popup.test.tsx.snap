// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Component should render correctly 1`] = `
<div
  class="sc-bdfBwQ lmGcYP"
  role="presentation"
>
  <div
    class="sc-gsTCUz bFFEVi"
  >
    <div
      class="sc-hKgILt"
    >
      <div
        class="sc-eCssSg"
      >
        Test header text
      </div>
      <button
        aria-label="Close popup"
        class="sc-jSgupP fcjYce"
        tabindex="0"
        type="button"
      >
        X
      </button>
    </div>
    <div
      class="sc-dlfnbm"
    >
      Test
    </div>
  </div>
</div>
`;
